src/components/ColorPicker.tsx

```diff
@@ -42,14 +47,61 @@ const ColorSwatch: FC<{
   const dispatch = useDispatch()
   const colors = useSelector(themeColors)
   const fontSize = useSelector(state => state.fontSize)
-  const selected =
-    (color && cursorStyle?.color === color) || (backgroundColor && cursorStyle?.backgroundColor === backgroundColor)
+  const currentThought = useSelector(state => !!state.cursor && getThoughtById(state, head(state.cursor))) as Thought
   size = size || fontSize * 1.2
 
+  /** A function that adds an alpha channel to a hex color. */
+  const addAlphaToHex = (hex: string) => {
+    if (hex.length === 7) return hex + 'ff'
+    return hex
+  }
+  const colorRegex = /color="(#[0-9a-fA-F]{6})"/g
+  const bgColorRegex = /background-color:\s*(rgb\(\d{1,3},\s?\d{1,3},\s?\d{1,3}\))/g
+  const selected = useMemo(() => {
+    const textHexColor = color ? addAlphaToHex(rgbToHex(color)) : undefined
+    const backHexColor = backgroundColor ? addAlphaToHex(rgbToHex(backgroundColor)) : undefined
+    const selectedTextHexColor = cursorStyle?.color ? addAlphaToHex(rgbToHex(cursorStyle?.color)) : undefined
+    const selectedBackHexColor = cursorStyle?.backgroundColor
+      ? addAlphaToHex(rgbToHex(cursorStyle?.backgroundColor))
+      : undefined
+    if ((cursorStyle?.color === undefined && cursorStyle?.backgroundColor === undefined) || !selection.isThought()) {
+      const currentThoughtValue = currentThought.value
+      const colorMatches = currentThoughtValue.match(colorRegex)
+      let matchColor, matchBgColor, match
+      const colors: Set<string> = new Set()
+      if (colorMatches) {
+        colorMatches.forEach(match => colors.add(match.slice(7, -1)))
+        matchColor = colors.size > 1 ? null : colors.values().next().value
+      }
+
+      const bgColors: Set<string> = new Set()
+      while ((match = bgColorRegex.exec(currentThoughtValue)) !== null) if (match[1]) bgColors.add(match[1])
+      matchBgColor = bgColors.size > 1 ? null : bgColors.values().next().value
+
+      return !!(
+        (textHexColor && textHexColor === (matchColor && addAlphaToHex(rgbToHex(matchColor)))) ||
+        (backHexColor && backHexColor === (matchBgColor && addAlphaToHex(rgbToHex(matchBgColor))))
+      )
+    }
+    return !!(
+      (textHexColor && textHexColor === selectedTextHexColor) ||
+      (backHexColor && backHexColor === selectedBackHexColor)
+    )
+  }, [cursorStyle, backgroundColor, color])
+
   /** Toggles the text color to the clicked swatch. */
   const toggleTextColor = (e: React.MouseEvent | React.TouchEvent) => {
-    // stop toolbar button dip
-    e.stopPropagation()
```

Comment:
Removing this is causing too many problems with the click behavior. Let's take a step back and consider other options. What are you trying to accomplish?